core/common/src/RawSource.kt

```diff
@@ -21,60 +21,31 @@
- * is both more efficient and more convenient. Use [buffer] to wrap any source with a buffer.
+ * The interface should be implemented to read data from wherever it's located: from the network, storage,
+ * or a buffer in memory. Sources may be layered to transform supplied data, such as to decompress, decrypt,
+ * or remove protocol framing.
  *
- * Sources are easy to test: just use a [Buffer] in your tests, and fill it with the data your
- * application is to read.
- *
- * ### Comparison with InputStream
-
- * This interface is functionally equivalent to [java.io.InputStream].
- *
- * `InputStream` requires multiple layers when consumed data is heterogeneous: a `DataInputStream`
- * for primitive values, a `BufferedInputStream` for buffering, and `InputStreamReader` for strings.
- * This library uses `BufferedSource` for all of the above.
- *
- * RawSource avoids the impossible-to-implement [available()][java.io.InputStream.available] method.
- * Instead callers specify how many bytes they [require][Source.require].
- *
- * RawSource omits the unsafe-to-compose [mark and reset][java.io.InputStream.mark] state that's
- * tracked by `InputStream`; instead, callers just buffer what they need.
- *
- * When implementing a source, you don't need to worry about the [read()][java.io.InputStream.read]
- * method that is awkward to implement efficiently and returns one of 257 possible values.
- *
- * And source has a stronger `skip` method: [Source.skip] won't return prematurely.
- *
- * ### Interop with InputStream
- *
- * Use [source] to adapt an `InputStream` to a source. Use [Source.inputStream] to adapt a
- * source to an `InputStream`.
+ * Most applications shouldn't operate on a raw source directly, but rather on a buffered [Source] which
+ * is both more efficient and more convenient. Use [buffer] to wrap any raw source with a buffer.
  */
-interface RawSource : Closeable {
-  /**
-   * Removes at least 1, and up to `byteCount` bytes from this and appends them to `sink`. Returns
-   * the number of bytes read, or -1 if this source is exhausted.
-   */
-  @Throws(IOException::class)
-  fun read(sink: Buffer, byteCount: Long): Long
-
+public interface RawSource : Closeable {
   /**
-   * Asynchronously cancel this source. Any [read] in flight should immediately fail with an
-   * [IOException], and any future read should also immediately fail with an [IOException].
+   * Removes at least 1, and up to [byteCount] bytes from this source and appends them to [sink].
+   * Returns the number of bytes read, or -1 if this source is exhausted.
+   *
+   * @param sink the destination to write the data from this source.
+   * @param byteCount the number of bytes to read.
    *
-   * Note that it is always necessary to call [close] on a source, even if it has been canceled.
+   * @throws IllegalArgumentException when [byteCount] is negative.
+   * @throws IllegalStateException when the source is closed.
    */
-  fun cancel()
```

Comment:
Worth creating an issue for this change to [re]evaluate it later

Reply:
I thought it should be added back with async support (although, there is no issue for that yet).

Reply:
It seemed like this method was useful even without explicit asynchrony (and was, in fact, not related to coroutines-based cancellation); I might be mistaken though, anyway worth initiating the discussion